### PR TITLE
Rustify Deno API

### PR DESCRIPTION
**Pull request by robbym**
Attempt to make the Rust Deno API more rusty.

Depends on #282

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -88,20 +90,22 @@ impl Deno {
...
         let source = CString::new(js_source).unwrap();
         let r = unsafe { deno_execute(self.ptr, filename.as_ptr(), source.as_ptr()) };
-        r != 0
-    }
-
-    fn last_exception(&self) -> &str {
-        let ptr = unsafe { deno_last_exception(self.ptr) };
-        let cstr = unsafe { CStr::from_ptr(ptr) };
-        cstr.to_str().unwrap()
+        if r != 0 {
```
Comment (matiasinsaurralde) [inline src/main.rs]: Is it possible to remove the additional block?
```rust
if r != 0 {
  Ok(())
}
let ptr = unsafe { deno_last_exception(self.ptr) };
let cstr = unsafe { CStr::from_ptr(ptr) };
Err(cstr.to_str().unwrap())
```

Comment (robbym) [reply src/main.rs]: Still needs to branch on the result of `deno_execute`.
Maybe something like:
```
    fn execute(&mut self, js_filename: &str, js_source: &str) -> Result<(), DenoException> {
        let filename = CString::new(js_filename).unwrap();
        let source = CString::new(js_source).unwrap();
        let r = unsafe { deno_execute(self.ptr, filename.as_ptr(), source.as_ptr()) };
        if r == 0 {
            let ptr = unsafe { deno_last_exception(self.ptr) };
            let cstr = unsafe { CStr::from_ptr(ptr) };
            Err(cstr.to_str().unwrap())
        }
        Ok(())
    }
```

Comment (matiasinsaurralde) [reply src/main.rs]: 👌

**Comment by ry** [comment]
Thank you and LGTM - as you can tell I am new to rust so the help is appreciated!
@piscisaureus please review.

**Comment by robbym** [comment]
@ry @matiasinsaurralde  I'm not completely satisfied with it, but I made a hypothetical change that prevents the wrong access path (using deno methods before calling init). https://github.com/robbym/deno/pull/1

Thoughts?

File: src/main.rs [hunk 2]
```diff
@@ -88,20 +90,21 @@ impl Deno {
         Deno { ptr: ptr }
     }
 
-    fn execute(&self, js_filename: &str, js_source: &str) -> bool {
+    fn execute(&mut self, js_filename: &str, js_source: &str) -> Result<(), DenoException> {
         let filename = CString::new(js_filename).unwrap();
```
Comment (F001) [inline src/main.rs]: I would like to suggest to introduce a real error type (an enum which encapsulates all the possible errors in the function), and get rid of `unwrap`s.

Comment (robbym) [reply src/main.rs]: Can do that tomorrow. Meanwhile please comment on https://github.com/robbym/deno/pull/1

**Comment by ry** [comment]
Included in #282 - which I will try to land soon - thanks @robbym !

**Pull request closed without merging**